src/Microsoft.Restier.Core/Query/QueryExpressionContext.cs

```diff
@@ -198,54 +266,114 @@ private QueryModelReference ComputeModelReference()
+                return ComputeParameterModelReference(parameter);
+            }
+
+            var member = this.VisitedNode as MemberExpression;
+            if (member != null)
+            {
+                return ComputeMemberModelReference(member);
+            }
+
+            return null;
+        }
+
+        private QueryModelReference ComputeParameterModelReference(ParameterExpression parameter)
+        {
+
+            QueryModelReference modelReference = null;
+            foreach (var node in this.GetExpressionTrail())
+            {
+                var methodCall = node as MethodCallExpression;
+                if (methodCall == null)
                 {
-                    methodCall = node as MethodCallExpression;
-                    if (methodCall == null)
-                    {
-                        continue;
-                    }
+                    continue;
+                }
+
+                modelReference = this.GetModelReferenceForNode(node);
+                if (modelReference == null)
+                {
+                    continue;
+                }
 
-                    modelReference = this.GetModelReferenceForNode(node);
-                    if (modelReference == null)
+                var method = methodCall.Method;
+                var sourceType = method.GetParameters()[0].ParameterType.FindGenericType(typeof(IEnumerable<>));
+                var resultType = method.ReturnType.FindGenericType(typeof(IEnumerable<>));
+                if (sourceType != resultType)
+                {
+                    // In case sourceType IEnumerable<Person> and resultType is IEnumerable<SelectExpandBinder.SelectAllAndExpand<Person>>
+                    // or IEnumerable<SelectExpandBinder.SelectAll<Person>> or IEnumerable<SelectExpandBinder.SelectSome<Person>>
+                    // or IEnumerable<SelectExpandBinder.SelectSomeAndInheritance<Person>>
+                    if (sourceType == null || resultType == null)
                     {
                         continue;
                     }
 
-                    var method = methodCall.Method;
-                    var sourceType = method.GetParameters()[0]
-                        .ParameterType.FindGenericType(typeof(IEnumerable<>));
-                    var resultType = method.ReturnType
-                        .FindGenericType(typeof(IEnumerable<>));
-                    if (sourceType != resultType)
+                    var resultGenericType = resultType.GenericTypeArguments[0];
+                    if (!resultGenericType.IsGenericType ||
+                        resultGenericType.GenericTypeArguments[0] != sourceType.GenericTypeArguments[0])
                     {
                         continue;
                     }
+                }
 
-                    var typeOfT = sourceType.GetGenericArguments()[0];
-                    if (parameter.Type == typeOfT)
+                var typeOfT = sourceType.GetGenericArguments()[0];
```

Comment:
what's the difference between `sourceType.GenericTypeArguments[0]` above and `sourceType.GetGenericArguments()[0]`?